File: moqt/track_reader.go
package moqt

import (
	"context"
	"errors"
	"sync"

	"github.com/okdaichi/gomoqt/quic"
)

func newTrackReader(broadcastPath BroadcastPath, trackName TrackName, subscribeStream *sendSubscribeStream, onCloseTrackFunc func()) *TrackReader {
	track := &TrackReader{
		BroadcastPath:       broadcastPath,
		TrackName:           trackName,
		sendSubscribeStream: subscribeStream,
		queuedCh:            make(chan struct{}, 1),
		queueing: make([]struct {
			sequence GroupSequence
			stream   quic.ReceiveStream
		}, 0, 1<<3),
		dequeued:         make(map[*GroupReader]struct{}),
		onCloseTrackFunc: onCloseTrackFunc,
	}

	return track
}

// TrackReader receives groups for a subscribed track.
// It queues incoming group streams and allows the application to accept them via AcceptGroup.
// TrackReader provides lifecycle and update APIs for managing subscriptions.
type TrackReader struct {
	BroadcastPath BroadcastPath
	TrackName     TrackName

	*sendSubscribeStream

	queueing []struct {
		sequence GroupSequence
		stream   quic.ReceiveStream
	}
	queuedCh chan struct{}
	trackMu  sync.Mutex

	dequeued map[*GroupReader]struct{}

	onCloseTrackFunc func()
}

// AcceptGroup blocks until the next group is available or context is
// canceled. It returns a GroupReader tied to the accepted group stream.
func (r *TrackReader) AcceptGroup(ctx context.Context) (*GroupReader, error) {
	trackCtx := r.Context()

	for {
		group := r.dequeueGroup()
		if group != nil {
			r.addGroup(group)

			return group, nil
		}

		if trackCtx.Err() != nil {
			return nil, Cause(trackCtx)
		}

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-trackCtx.Done():
			return nil, Cause(trackCtx)
		case <-r.queuedCh:
		}
	}
}

func (r *TrackReader) dequeueGroup() *GroupReader {
	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	if len(r.queueing) > 0 {
		next := r.queueing[0]

		r.queueing = r.queueing[1:]

		var group *GroupReader
		group = newGroupReader(next.sequence, next.stream,
			func() { r.removeGroup(group) })

		return group
	}

	return nil
}

// Close cancels queued groups, closes the queued channel, and terminates
// the subscription stream gracefully.
func (r *TrackReader) Close() error {
	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	// Cancel all pending groups first
	errCode := quic.StreamErrorCode(SubscribeCanceledErrorCode)
	for _, entry := range r.queueing {
		entry.stream.CancelRead(errCode)
	}
	r.queueing = nil

	// Cancel all dequeued groups
	for stream := range r.dequeued {
		stream.CancelRead(SubscribeCanceledErrorCode)
	}
	r.dequeued = nil

	if r.queuedCh != nil {
		close(r.queuedCh)
		r.queuedCh = nil
	}

	r.onCloseTrackFunc()

	return r.sendSubscribeStream.close()
}

// CloseWithError cancels the subscription with the provided SubscribeErrorCode and terminates the subscription.
func (r *TrackReader) CloseWithError(code SubscribeErrorCode) error {
	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	// Cancel all pending groups first
	errCode := quic.StreamErrorCode(code)
	for _, entry := range r.queueing {
		entry.stream.CancelRead(errCode)
	}
	r.queueing = nil

	// Cancel all dequeued groups
	for stream := range r.dequeued {
		stream.CancelRead(SubscribeCanceledErrorCode)
	}
	r.dequeued = nil

	if r.queuedCh != nil {
		close(r.queuedCh)
		r.queuedCh = nil
	}

	r.onCloseTrackFunc()

	return r.sendSubscribeStream.closeWithError(code)
}

// Update updates the subscription configuration with a new TrackConfig.
func (r *TrackReader) Update(config *TrackConfig) error {
	if config == nil {
		return errors.New("subscribe config cannot be nil")
	}

	return r.sendSubscribeStream.updateSubscribe(config)
}

// TrackConfig returns the currently active subscription configuration.
func (r *TrackReader) TrackConfig() *TrackConfig {
	return r.sendSubscribeStream.TrackConfig()
}

func (r *TrackReader) enqueueGroup(sequence GroupSequence, stream quic.ReceiveStream) {
	if stream == nil {
		return
	}

	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	if r.Context().Err() != nil || r.queueing == nil {
		stream.CancelRead(quic.StreamErrorCode(SubscribeCanceledErrorCode))
		return
	}

	entry := struct {
		sequence GroupSequence
		stream   quic.ReceiveStream
	}{
		sequence: sequence,
		stream:   stream,
	}
	r.queueing = append(r.queueing, entry)

	select {
	case r.queuedCh <- struct{}{}:
	default:
	}
}

func (r *TrackReader) addGroup(group *GroupReader) {
	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	r.dequeued[group] = struct{}{}
}

func (r *TrackReader) removeGroup(group *GroupReader) {
	r.trackMu.Lock()
	defer r.trackMu.Unlock()

	delete(r.dequeued, group)
}

File: moq-web/src/track_reader.ts
import { GroupReader } from "./group_stream.ts";
import type { Info } from "./info.ts";
import type { Context } from "@okudai/golikejs/context";
import { ContextCancelledError, watchPromise } from "@okudai/golikejs/context";
import type { SendSubscribeStream, TrackConfig } from "./subscribe_stream.ts";
import type { ReceiveStream } from "./internal/webtransport/mod.ts";
import { GroupMessage } from "./internal/message/mod.ts";
import type { BroadcastPath } from "./broadcast_path.ts";
import { Queue } from "./internal/queue.ts";

export class TrackReader {
  broadcastPath: BroadcastPath;
  trackName: string;
  #subscribeStream: SendSubscribeStream;
  #queue: Queue<[ReceiveStream, GroupMessage]>;
  #onCloseFunc: () => void;

  constructor(
    broadcastPath: BroadcastPath,
    trackName: string,
    subscribeStream: SendSubscribeStream,
    queue: Queue<[ReceiveStream, GroupMessage]>,
    onCloseFunc: () => void,
  ) {
    this.broadcastPath = broadcastPath;
    this.trackName = trackName;
    this.#subscribeStream = subscribeStream;
    this.#queue = queue;
    this.#onCloseFunc = onCloseFunc;
  }

  async acceptGroup(
    signal: Promise<void>,
  ): Promise<[GroupReader, undefined] | [undefined, Error]> {
    // Check if context is already cancelled
    const err = this.context.err();
    if (err) {
      return [undefined, err];
    }

    while (true) {
      const ctx = watchPromise(this.context, signal);
      const dequeued = await Promise.race([
        this.#queue.dequeue(),
        ctx.done().then(() => {
          return new ContextCancelledError() as Error;
        }),
        this.context.done().then(() => {
          return new Error(
            `track reader context cancelled: ${this.context.err()?.message}`,
          );
        }),
      ]);

      if (dequeued instanceof Error) {
        return [undefined, dequeued];
      }
      if (dequeued === undefined) {
        // This is
        throw new Error("dequeue returned undefined");
      }

      const [reader, msg] = dequeued;

      const group = new GroupReader(this.context, reader, msg);

      return [group, undefined];
    }
  }

  async update(config: TrackConfig): Promise<Error | undefined> {
    return this.#subscribeStream.update(config);
  }

  readInfo(): Info {
    return this.#subscribeStream.info;
  }

  async closeWithError(code: number): Promise<void> {
    await this.#subscribeStream.closeWithError(code);
    this.#onCloseFunc();
  }

  get trackConfig(): TrackConfig {
    return this.#subscribeStream.config;
  }

  get context(): Context {
    return this.#subscribeStream.context;
  }
}
